src/tracer.c

```diff
@@ -75,210 +75,189 @@ next_finfo(PROCESS_INFO *pi)
     return &(pi->finfo[++(pi->numfinfo)]);
 }
 
-char *
+char           *
 get_str_from_process(pid_t pid, void *addr)
 {
-    static char buf[PATH_MAX];
-    char *dest = buf;
-    union
-    {
-	long lval;
-	char cval[sizeof (long)];
+    static char     buf[PATH_MAX];
+    char           *dest = buf;
+    union {
+	long            lval;
+	char            cval[sizeof (long)];
     } data;
 
-    for (int i = 0;; i++)
-    {
+    size_t          i = 0;
+
+    do {
 	data.lval =
 		ptrace(PTRACE_PEEKDATA, pid, addr + i * sizeof (long), NULL);
-	for (int j = 0; j < sizeof (long); j++)
-	{
+	for (int j = 0; j < sizeof (long); j++) {
 	    *dest++ = data.cval[j];
 	    if (data.cval[j] == 0)
 		break;
 	}
-    }
+	++i;
+    } while (*dest);
     return strdup(buf);
 }
 
-/*====================================================================================*/
-
-#if 0
-// still TODO
+#define find(arr, size, pid)						\
+({ 									\
+    size_t i = 0;							\
+    for (; i < size; ++i)						\
+    {									\
+	if (arr[i].pid == pid)						\
+	{								\
+	    break;							\
+	}								\
+    }									\
+									\
+									\
+    if(i == size)							\
+    { 									\
+        error(EXIT_FAILURE, errno, "process %d isn't in array\n", pid);	\
+    }									\
+    arr + i; 								\
+})
 
 static void
 handle_syscall(pid_t pid, const struct ptrace_syscall_info *entry,
-	       const struct ptrace_syscall_info *exit, files * buffer)
+	       const struct ptrace_syscall_info *exit)
 {
-    if (exit->exit.rval < 0)
-    {
+    if (exit->exit.rval < 0) {
 	return;			       // return on syscall failure
     }
 
-    int syscall = entry->entry.nr;
+    FILE_INFO      *finfo = next_finfo(find(pinfo, numpinfo + 1, pid));
```

Comment:
You shouldn't always get a new FILE_INFO:
- if it's `open`, record the filename and save it at the `fd` place of the array;
- if it's `close`, based on the `fd` get the entry and `process_closed_file` (get the contents)

Reply:
@zvr I've thought about it, i plan on having an extra 'prologue' `if` statement taking care of the `SYS_CLOSE` scenario. I didn't implement it here because this branch is supposed to only rewrite stuff we already wrote. Tell me if otherwise.

Reply:
@zvr looking at it again, we have a problem.

finfo index starts counting normally from zero, but `stdin`, `stdout`, `stderr` already reserve the file descriptor numbers 0, 1 and 2.

Yesterday i suggested we use `fd - 3` in order to ignore those three, they are not directly opened from the program and i struggle to find any reason why we should care about them.

So we either do that or we now initialize `numinfo` with `2` instead of `-1`.